packages/react-query-next-experimental/src/HydrationStreamProvider.tsx

```diff
@@ -0,0 +1,194 @@
+     * Example devalue, superjson et al
+     */
+    transformer?: DataTransformer
+    /**
+     * **Client method**
+     * Called in the browser when new entries are received
+     */
+    onEntries: (entries: TShape[]) => void
+    /**
+     * **Server method**
+     * onFlush is called on the server when the cache is flushed
+     */
+    onFlush?: () => TShape[]
+  }) {
+    // unique id for the cache provider
+    const id = `_${React.useId()}`
+    const idJSON = JSON.stringify(id)
+
+    const [transformer] = React.useState(
+      () =>
+        (props.transformer ?? {
+          // noop
+          serialize: (obj: any) => obj,
+          deserialize: (obj: any) => obj,
+        }) as unknown as TypedDataTransformer<TShape>,
+    )
+
+    // <server stuff>
+    const [stream] = React.useState<TShape[]>(() => {
+      if (typeof window !== 'undefined') {
+        return {
+          push() {
+            // no-op on the client
+          },
+        } as unknown as TShape[]
+      }
+      return []
+    })
+    const count = React.useRef(0)
+    const onDehydrateRef = React.useRef(props.onFlush)
+    onDehydrateRef.current = props.onFlush
+    useServerInsertedHTML(() => {
+      // This only happens on the server
+      stream.push(...(onDehydrateRef.current?.() ?? []))
+
+      if (!stream.length) {
+        return null
+      }
+      // console.log(`pushing ${stream.length} entries`)
+      const serializedCacheArgs = stream
+        .map((entry) => transformer.serialize(entry))
+        .map((entry) => JSON.stringify(entry))
+        .join(',')
+
+      // Flush stream
+      stream.length = 0
+
+      const html: string[] = [
+        `window[${idJSON}] = window[${idJSON}] || [];`,
+        `window[${idJSON}].push(${serializedCacheArgs});`,
+      ]
+      return (
+        <script
+          key={count.current++}
+          dangerouslySetInnerHTML={{
+            __html: html.join(''),
+          }}
+        />
+      )
+    })
+    // </server stuff>
+
+    // <client stuff>
+    const onEntriesRef = React.useRef(props.onEntries)
+    onEntriesRef.current = props.onEntries
+
+    // Client: consume cache:
+    const onEntries = React.useCallback(
+      (...serializedEntries: Serialized<TShape>[]) => {
+        const entries = serializedEntries.map((serialized) =>
+          transformer.deserialize(serialized),
+        )
+        onEntriesRef.current(entries)
+      },
+      [transformer],
+    )
+
+    React.useEffect(() => {
+      const win = window as any
+      // Register cache consumer
+      const stream: Array<Serialized<TShape>> = win[id] ?? []
+
+      if (!Array.isArray(stream)) {
+        throw new Error(`${id} seem to have been registered twice`)
+      }
+      onEntries(...stream)
+
+      // Register our own consumer
+      win[id] = {
+        push: onEntries,
+      }
+
+      return () => {
+        // Cleanup after unmount
+        win[id] = {
```

Comment:
This should probably be changed to an empty array instead

Reply:
okay: a83ca67b